cmd/format/format.go

```diff
@@ -344,21 +352,30 @@ func applyFormatters(
 				}
 			}
 
-			// see if any formatters matched
+			// check for no matches
 			if len(matches) == 0 {
-
+				// log that there was no match, exiting with an error if the unmatched level was set to fatal
 				if unmatchedLevel == log.FatalLevel {
 					return fmt.Errorf("no formatter for path: %s", file.RelPath)
 				}
+
 				log.Logf(unmatchedLevel, "no formatter for path: %s", file.RelPath)
-				// mark it as processed and continue to the next
-				formattedCh <- &format.Task{
-					File: file,
+
+				// release the file
+				if err := file.Release(); err != nil {
+					return fmt.Errorf("failed to release file: %w", err)
 				}
```

Comment:
This looks a lot like the file release code up above. Is there some way we can restructure this `for` loop so we can ensure the file is always released, regardless of the codepath? Could go's `defer` statement help here?

(for the record, I don't know what "releasing" a file means, nor why it's something we didn't do before but have to do now)

Reply:
`defer` is bound to a function and can't be used within a `for` loop. I'll try restructuring the logic to make it clearer and require only one release block. 

FYI: I'm playing with the concept of releasing in the layered `walk.Reader`. Any reader implementation can append a hook that gets run when `release` is called. Currently, the cached reader is using that to update the cache, and the stdin reader uses it to delete the temporary file it created.

Reply:
I restructured it like this and actually caught a missing release in the process:

https://github.com/numtide/treefmt/blob/52ac241979fd34dc4f3b0a489977bb928d184ade/cmd/format/format.go#L334-L383

Reply:
Cool! The restructuring looks good to me, thanks. Glad it caught a bug!

My poor python brain is distressed to see cleanup code written so linearly. I'm very hard coded to assume that any line of code can throw an exception, and I need to use a `try...finally` or a `with` to ensure things actually happen. I guess that's not necessary in go because there are no implicit goto exceptions, just explicit required error handling?

For the record, it's still hard for me to convince myself we actually always release these files, because I don't see it happening here for the files we matched. Perhaps there's a slightly different structure to all of this that would make even that clear?

(don't take this as actionable feedback, I'm just thinking out loud).

Reply:
After I merge the in-flight PRs, I'll update the `golangci-lint` config to match https://github.com/nix-community/go-nix and do a general cleanup. Then, I want to focus on making the workflow easier to follow, as, like you, I'm not 100% comfortable with it right now.